cql3/statements/property_definitions.cc

```diff
@@ -162,14 +189,70 @@ void property_definitions::remove_from_map_if_exists(const sstring& name, const
         return;
     }
     try {
-        auto map = std::get<map_type>(it->second);
+        auto map = std::get<extended_map_type>(it->second);
         map.erase(key);
         _properties[name] = map;
     } catch (const std::bad_variant_access& e) {
         throw exceptions::syntax_exception(format("Invalid value for property '{}'. It should be a map.", name));
     }
 }
 
+/// Converts extended map into a flat map.
+///
+/// Values which are lists are represented as multiple entries in the map
+/// with the list index appended to the key, with ':' as a separator.
+///
+/// For example:
+///
+///    {'dc1': '3', 'dc2': ['rack1', 'rack2']}
+///
+/// has a flattened representation of:
+///
+///   {'dc1': '3', 'dc2:0': 'rack1', 'dc2:1': 'rack2'}
+///
+property_definitions::map_type to_flat_map(const property_definitions::extended_map_type& in) {
+    property_definitions::map_type out;
+    for (const auto& [in_key, in_value]: in) {
+        std::visit(overloaded_functor{
+            [&] (const sstring& value) {
+                out[in_key] = value;
+            },
+            [&] (const std::vector<sstring>& list) {
+                if (list.empty()) {
+                    throw_with_backtrace<std::runtime_error>(fmt::format("empty list for key '{}'", in_key));
+                } else {
+                    // flatten the rack list in multiple entries
+                    for (size_t i = 0; i < list.size(); ++i) {
+                        const auto& v = list[i];
+                        out[in_key + ":" + std::to_string(i)] = v;
+                    }
+                }
+            }
+        }, in_value);
+    }
+    return out;
+}
+
+property_definitions::extended_map_type from_flat_map(const property_definitions::map_type& in) {
+    property_definitions::extended_map_type out;
+    for (const auto& [key, value] : in) {
+        auto pos = key.find(':');
```

Comment:
let's use `find_last_of` instead.
and add a FIXME to validate datacenter name in case it contains a `:` character.

Reply:
Even better is to add validation to `to_flat_map()`:
```
        if (in_key.find(':') != sstring::npos) {
            throw std::invalid_argument(fmt::format("key '{}' contains reserved character ':'", in_key));
        }
```